ui/vty-lib/src/Ariadne/UI/Vty/Widget/Repl.hs

```diff
@@ -219,10 +235,50 @@ handleReplWidgetKey = \case
     historyNavigate action = do
       ReplWidgetState{..} <- get
       cmd <- liftIO $ action replWidgetHistoryFace
-      whenJust cmd $ assign replWidgetCommandL
+      whenJust cmd $ \cmd' -> do
+        let
+          ls = T.splitOn "\n" cmd'
+          newLoc = (length ls - 1, T.length $ ls !! (length ls - 1))
+        assign replWidgetCommandL cmd'
+        assign replWidgetCommandLocationL newLoc
       reparse
       return WidgetEventHandled
 
+handleAutocompleteKey
+  :: (forall a. (Z.Zipper a -> Z.Zipper a))
+  -> WidgetEventM (ReplWidgetState p) p WidgetEventResult
+handleAutocompleteKey move = do
+  ReplWidgetState{..} <- get
+  options <- case replWidgetCurrentAutocompletion of
+    Nothing -> do
+      let
+        ls = T.splitOn "\n" replWidgetCommand
+        (locRow, locColumn) = replWidgetCommandLocation
+        cmdBeforeLocation = T.intercalate "\n"
+          $ snoc (take locRow ls) (T.take locColumn (ls !! locRow))
+        cmdAfterLocation = T.intercalate "\n"
+          $ T.drop locColumn (ls !! locRow) : drop (locRow + 1) ls
+        prefixSuggestions = langAutocomplete replWidgetLangFace cmdBeforeLocation
+        suggestions = (`map` prefixSuggestions) $ \suggestion ->
+          let ls' = T.splitOn "\n" suggestion
+          in (suggestion <> cmdAfterLocation, (length ls' - 1, length $ Universum.Unsafe.last ls'))
```

Comment:
You imported `Universum.Unsafe.last` and used it only here, but in other places (including one in `Ariadne.UI.Vty.Widget.Form.Edit`, and always to get the length of the last line) you use `T.length $ ls !! (length ls - 1)` that AFAIK has a higher cost (traverses the list twice: to get the length, and then to apply `!!`) and causes repetition.

Also you usually use `T.splitOn "\n"` and `T.intercalate "\n"` to divide and recombine lines, when there are functions in `Data.Text` to do exactly this.

What I suggest is:
- use `T.lines` instead of `T.splitOn "\n"`(NOTE: there is actually a difference between the two: `T.lines` will give back an empty list for a empty input, instead of a `[""]`, so you might need to change a bit of code to adjust the logic in `handleAutocompleteKey`)
- use `T.unlines` instead of `T.intercalate "\n"`
- write a function (for the `Edit` widget too) that calculates the position tuple out of a `Text`, that checks if it's empty (see 2 point above) and as such is in fact not unsafe anymore

I hope I explained myself well enough.

P.s. you can see from this very line that you can use `length` on `Text`, because `Universum`'s `Containers`' `length` has a `Data.Text` implementation (that uses `T.length`).

Reply:
`T.lines " \n"` also evaluates to `[" "]` while we need it to be `[" ",""]`. I suppose that that small difference between `T.lines` and `T.splitOn "\n"` was precisely the reason the latter was used in `Ariadne.UI.Vty.Widget.Form.Edit`, I just followed the example. Using `T.lines` and `T.unlines` will just add a lot of `if`s.

Agreed with all other points.

Reply:
Ok, if that's the case keep using them

Reply:
I think I'll leave position calculation in `Edit` as is because the scenario is slightly different there and just get rid of copy-paste in `Repl`.